1. What is the difference between a function and a method in Python?

Ans. In Python, both functions and methods are blocks of code that perform specific tasks, but they differ primarily in how they are called and where they are defined. Here’s a breakdown of the differences:

1. Functions:
Definition: A function is defined using the def keyword and can exist independently of any object.
Usage: Functions are called by their name followed by parentheses. They can take arguments and return values.
Scope: Functions can be defined at the top level of a module, inside other functions, or even inside methods.

2. Methods:
Definition: A method is a function that is associated with an object and defined within a class. Methods are functions that operate on instances of classes.
Usage: Methods are called on objects (instances of classes) using the dot notation (object.method()).
Scope: Methods are always defined inside a class and typically take self as their first parameter, which refers to the instance of the class.

Key Differences:
Association: Methods are associated with objects (class instances), while functions are standalone.
Calling: Functions are called directly by their name, whereas methods are called on objects using the dot notation.
First Parameter: Methods typically have self as their first parameter to access instance-specific data, while functions don’t have this requirement.
These distinctions are crucial when designing and organizing code in object-oriented programming 

2. Explain the concept of function arguments and parameters in Python.

Ans. In Python, the terms arguments and parameters are often used interchangeably, but they have distinct meanings in the context of functions. Here’s a breakdown of each concept:

1. Parameters:
Definition: Parameters are the variables listed inside the parentheses in the function definition. They act as placeholders for the values that will be passed into the function when it is called.
Role: Parameters allow you to define functions that can accept input and perform operations based on that input.

2. Arguments:
Definition: Arguments are the actual values that are passed to the function when it is called. They correspond to the parameters defined in the function.
Role: Arguments provide the actual data that the function will use to perform its operations.

Key Differences:
Location:

Parameters are part of the function definition.
Arguments are part of the function call.
Function Definition vs. Function Call:

When you define a function, you specify parameters.
When you invoke or call the function, you pass in arguments.

3. What are the different ways to define and call a function in Python?

Ans. In Python, functions can be defined and called in various ways, depending on the use case. Here's an overview of the different ways to define and call functions:

1. Regular Function Definition and Call
Definition:
You define a function using the def keyword, followed by the function name and parentheses containing any parameters.

2. Function with Default Arguments
Definition:
You can define a function with default parameter values, which are used if no argument is provided for those parameters.

3. Function with Keyword Arguments
Definition:
Same as a regular function, but you can call it using keyword arguments to explicitly specify which parameter each argument is for.

4. Function with Variable-Length Arguments
Definition:
You can define a function that accepts a variable number of positional (*args) or keyword arguments (**kwargs).

4. What is the purpose of the `return` statement in a Python function?

Ans. The return statement in a Python function is used to exit the function and send a value (or values) back to the caller. It serves several important purposes:

1. Returning a Value:
The primary purpose of the return statement is to provide a way for the function to output a result that can be used elsewhere in the code.
When a function is called, the return statement provides the output that replaces the function call.

2. Exiting the Function:
The return statement also causes the function to terminate immediately. Any code written after the return statement within the function will not be executed.

5. What are iterators in Python and how do they differ from iterables?

Ans. In Python, iterators and iterables are related concepts that are central to handling sequences of data, particularly in loops. Here's a breakdown of what each term means and how they differ:

1. Iterables:
Definition:
An iterable is any Python object that can return an iterator. It is an object that can be iterated (looped) over.
Examples of iterables include sequences like lists, tuples, strings, and also non-sequence collections like dictionaries and sets.
Characteristics:
__iter__() method: An object is considered iterable if it implements the __iter__() method, which returns an iterator.

2. Iterators:
Definition:
An iterator is an object that represents a stream of data. It returns one element at a time when you call its __next__() method (or next() function in Python 2).
An iterator is created from an iterable using the iter() function.
Characteristics:
__next__() method: An iterator must implement the __next__() method, which returns the next item in the sequence. When there are no more items to return, it raises a StopIteration exception.
Stateful: Iterators maintain internal state, keeping track of where they are in the sequence.

3. Key Differences Between Iterables and Iterators:
Definition:

An iterable is any object that can be iterated over (like lists, tuples, and strings).
An iterator is an object that does the actual iteration over the iterable.
Method Implementation:

An iterable has an __iter__() method that returns an iterator.
An iterator has both __iter__() (which returns itself) and __next__() methods.
Usage:

Iterables are typically used to represent collections of items, while iterators are used to traverse through these collections.
When you use a for loop on an iterable, Python automatically calls iter() on the iterable to get an iterator, and then it repeatedly calls next() on the iterator to retrieve each element.
State:

Iterables do not maintain any state about the iteration. They simply provide a fresh iterator each time __iter__() is called.
Iterators maintain the state of where they are in the sequence, so you can think of them as a pointer moving through the iterable.

6. Explain the concept of generators in Python and how they are defined.

Ans. Generators in Python are a special type of iterable, similar to iterators, that allow you to iterate over a sequence of values lazily, meaning they generate values on the fly and do not store the entire sequence in memory at once. This makes generators very memory-efficient, especially when working with large datasets.

Key Concepts of Generators:
Lazy Evaluation:

Generators produce items one at a time only when requested, which is known as lazy evaluation. This is in contrast to lists, where all elements are stored in memory.
This is particularly useful when dealing with large datasets or streams of data that don’t need to be fully loaded into memory.
Yield Keyword:

Generators are defined using a regular function but use the yield keyword instead of return to produce a value and pause the function’s execution. The function's state is preserved between each call to next() on the generator.
When yield is encountered, the function’s state is saved, and the yielded value is returned to the caller. The function can be resumed from where it left off the next time it’s called.
Memory Efficiency:

Because generators yield items one at a time, they are much more memory-efficient than using a list to hold all items in memory at once.

7. What are the advantages of using generators over regular functions?

Ans. Generators offer several advantages over regular functions, particularly in scenarios involving large data processing, memory management, and streamlining complex iterations. Here are the key advantages of using generators:

1. Memory Efficiency:
Lazy Evaluation: Generators produce items one at a time and only when needed, instead of generating and storing all values at once like a list or other collections. This lazy evaluation makes generators more memory-efficient, especially when dealing with large datasets or infinite sequences.
No Need for Entire Data Storage: Since generators don’t need to store all items in memory, they are ideal for working with large data that can’t fit in memory all at once.

2. Improved Performance:
Faster Execution: Because generators yield values as needed, they can start producing output more quickly, making them faster in scenarios where not all data is required at once. This is in contrast to functions that return lists, where all processing must complete before any data is returned.
Efficient Looping: In loops, generators can improve performance by providing values on-the-fly, reducing the time spent on generating and storing intermediate results.

3. Handling Infinite Sequences:
Support for Infinite Iteration: Generators are well-suited for generating infinite sequences where generating all elements upfront is impossible. Regular functions cannot handle infinite sequences since they would attempt to create an infinite list, leading to memory exhaustion.

8. What is a lambda function in Python and when is it typically used?

Ans. A lambda function in Python is a small, anonymous function defined using the lambda keyword. It can have any number of arguments but only one expression. The expression is evaluated and returned when the function is called. Lambda functions are often used for short, simple operations where defining a full function using def would be unnecessarily verbose.

Syntax of Lambda Functions:
The basic syntax of a lambda function is:
lambda: This keyword is used to create a lambda function.
arguments: These are the inputs to the function, similar to the parameters in a regular function.
expression: This is a single expression that the lambda function evaluates and returns.


Typical Uses of Lambda Functions:
Lambda functions are typically used in situations where a small, simple function is required for a short period, often as an argument to higher-order functions like map(), filter(), and sorted(). Here are some common use cases:

1. As Arguments to Higher-Order Functions:
map() Function:
The map() function applies a given function to all items in an iterable (e.g., a list) and returns a map object (which is an iterator).

filter() Function:
The filter() function filters elements in an iterable based on a given function that returns True or False.

Limitations of Lambda Functions:
Single Expression Only: Lambda functions can only contain a single expression. This limits their use to simple operations.
No Statements: Lambda functions cannot contain statements such as loops, if, else, try, except, etc.
Less Readable for Complex Logic: While lambda functions are concise, they can make the code harder to read and understand if used for complex logic.

9. Explain the purpose and usage of the `map()` function in Python.

Ans. The map() function in Python is a built-in function that applies a given function to each item in an iterable (like a list, tuple, or string) and returns a map object (which is an iterator) containing the results. It's a powerful tool for transforming data by applying a specific operation across all elements of a collection without writing an explicit loop.

Syntax of map(): 
function: A function that defines the operation to apply to each element in the iterable(s). This function can be a built-in function, a user-defined function, or a lambda function.
iterable: One or more iterable(s) (e.g., lists, tuples, strings) whose elements you want to process with the function.


How map() Works:
The map() function applies the function to each element of the provided iterable(s) and returns an iterator of the results.
If multiple iterables are provided, function should accept the same number of arguments as there are iterables. map() will then apply the function to the corresponding elements of each iterable.

10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

Ans. The map(), reduce(), and filter() functions in Python are higher-order functions that operate on iterables (like lists, tuples, etc.) and allow for functional-style programming. Each of these functions has a distinct purpose and behavior.

1. map() Function:
Purpose: The map() function applies a given function to each item in an iterable (or iterables) and returns an iterator that yields the results.

Operation: It transforms each element in the iterable(s) by applying the provided function.

2. reduce() Function:
Purpose: The reduce() function, from the functools module, applies a binary function (a function that takes two arguments) cumulatively to the items of an iterable, reducing the iterable to a single value.

Operation: It processes the iterable by applying the function cumulatively, typically for operations like summing, multiplying, or concatenating items.

3. filter() Function:
Purpose: The filter() function applies a predicate function (a function that returns a boolean) to each item in an iterable and returns an iterator yielding only the elements for which the predicate is True.

Operation: It filters the elements in the iterable based on the condition defined in the predicate function.

11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
list:[47,11,42,13];

In [13]:
from functools import reduce

lis = [47,11,42,13]
sum_lis= reduce(lambda x, y:x+y, lis)
print(sum_lis)

113


in above code lis is iterable and 'lambda x, y:x+y' is function. reduce function executed lambda function for each pair of iterable lis.
for first iteration 
x=47
y=11
x+y=58
for second iteration
x=58
y=42
x+y=100
for third iteration
x=100
y=13
x+y=113

finally there is no more digit so output will be 113.

12. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in
the list.

In [14]:
def even_sum(list):
    total=0
    for i in list:
        if i%2==0:
            total=total+i

    return total

13. Create a Python function that accepts a string and returns the reverse of that string.

In [16]:
def rev_str(strings):
    x=strings[::-1]
    return x

14.  Implement a Python function that takes a list of integers and returns a new list containing the squares of
each number.

In [18]:
def squar(lis):
    x=[i*i for i in lis]
    return x

15. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [32]:
def prime_number(num):
    a=1
    for i in range (num-2):
        if num%(num-a)==0:
            x=1
            break
        else:
            x=0
        a+=1
    return x 

def check_prime(number):
    

    if prime_number(number)==0:
        print(f"{number} is a prime number")
    else:
        print(f"{number} isn't a prime number")


16. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of
terms.

In [48]:
def fibonacci_series(number):
    z=[0,1]
    a=0
    b=1
    for i in range (2, number):
        
        fit=a+b
        z.append(fit)
        yield z
        a=b
        b=a+b


17.  Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [53]:
def power_func(number):
    for i in range(1, number+1):
        yield 2**i

In [ ]:
18.  Implement a generator function that reads a file line by line and yields each line as a string.

In [ ]:
def read_file_line_by_line(file_path):
    with open(file_path, 'r') as file:  # Open the file in read mode
        for line in file:  # Iterate over each line in the file
            yield line.strip()  # Yield the line after stripping whitespace

# Example usage:
file_path = 'example.txt'  # Replace with your file path
for line in read_file_line_by_line(file_path):
    print(line)  # Process each line as needed


19. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [ ]:
# List of tuples
data = [(1, 'banana'), (2, 'apple'), (3, 'cherry'), (4, 'date')]

# Sort the list of tuples based on the second element of each tuple
sorted_data = sorted(data, key=lambda x: x[1])

# Print the sorted list
print(sorted_data)  # Output: [(2, 'apple'), (4, 'date'), (1, 'banana'), (3, 'cherry')]


20. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [59]:
temp=[20, 25, 22,27,35,40]
far=list(map(lambda x:(x*9/5)+32, temp))
print(far)

[68.0, 77.0, 71.6, 80.6, 95.0, 104.0]


21. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [79]:
a='ayushmankashyap'

def vowel(i):
    strings=''
    if i!='a' and i!='e' and i!='i' and i!='o' and i!='u':
        strings= i+strings
    else:
        strings=strings

    return strings

x= filter (vowel, a )
#x=list(x)
x=''.join(x)

print(x)

yshmnkshyp


22.  Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:







Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the
product of the price per item and the quantity. The product should be increased by 10,- € if the value of the
order is smaller than 100,00 €.

Write a Python program using lambda and map.

In [ ]:
# Sample data representing orders (order_number, price_per_item, quantity)
orders = [
    [1, 15.00, 2],  # Order 1: 2 items at €15 each
    [2, 30.00, 3],  # Order 2: 3 items at €30 each
    [3, 50.00, 1],  # Order 3: 1 item at €50
    [4, 5.00, 5],   # Order 4: 5 items at €5 each
    [5, 100.00, 1]  # Order 5: 1 item at €100
]

# Function to calculate order value
def calculate_order_value(order):
    order_number, price_per_item, quantity = order
    total_value = price_per_item * quantity
    # Add €10 if total value is less than €100
    if total_value < 100:
        total_value += 10
    return (order_number, total_value)

# Use map to apply the function to each order
result = list(map(calculate_order_value, orders))

# Print the result
print(result)
